src/main/java/com/lovetropics/minigames/common/core/game/behavior/action/GameActionList.java

```diff
@@ -13,92 +12,103 @@
 import com.mojang.serialization.MapCodec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
 import net.minecraft.server.level.ServerPlayer;
-import net.minecraft.util.StringRepresentable;
+import net.minecraft.util.ExtraCodecs;
 
 import java.util.Arrays;
 import java.util.List;
 import java.util.function.Function;
+import java.util.function.Supplier;
+
+public class GameActionList<T> {
+    public static final GameActionList<ServerPlayer> EMPTY = new GameActionList<>(List.of(), PlayerActionTarget.SOURCE);
+
+    public static final MapCodec<GameActionList<?>> MAP_CODEC = RecordCodecBuilder.mapCodec(i -> i.group(
+            MoreCodecs.listOrUnit(IGameBehavior.CODEC).fieldOf("actions").forGetter(list -> list.behaviors),
+            ActionTarget.FALLBACK_PLAYER.optionalFieldOf("target", PlayerActionTarget.SOURCE).forGetter(list -> list.target)
+    ).apply(i, GameActionList::new));
+
+    private static final Codec<GameActionList<?>> SIMPLE_CODEC = MoreCodecs.listOrUnit(IGameBehavior.CODEC)
+            .flatComapMap(
+                    behaviors -> new GameActionList<>(behaviors, PlayerActionTarget.SOURCE),
+                    list -> {
+                        if (!(list.target instanceof PlayerActionTarget tg) || tg.target() != PlayerActionTarget.Target.SOURCE) {
+                            return DataResult.error(() -> "Cannot encode simple action list with target: " + list.target);
+                        }
+                        return DataResult.success(list.behaviors);
+                    }
+            );
+
+    public static final Codec<GameActionList<?>> TYPE_SAFE_CODEC = Codec.either(SIMPLE_CODEC, MAP_CODEC.codec())
+            .xmap(either -> either.map(Function.identity(), Function.identity()), Either::right);
+
+    public static final Codec<GameActionList<ServerPlayer>> PLAYER = mandateType(ActionTargetTypes.PLAYER);
```

Comment:
This can be `mandateTypeDefaultingTarget(ActionTargetTypes.PLAYER, PlayerActionTarget.SOURCE)` right? In that case we might be able to drop `SIMPLE_CODEC` and `TYPE_SAFE_CODEC`, too